### PR TITLE
[SPARK-45054][SQL] HiveExternalCatalog.listPartitions should restore partition statistics

### PR DESCRIPTION
**Pull request by sunchao**
### What changes were proposed in this pull request?

Call `restorePartitionMetadata` in `listPartitions` to restore Spark SQL statistics.

### Why are the changes needed?

Currently when `listPartitions` is called, it doesn't restore Spark SQL statistics stored in metastore, such as `spark.sql.statistics.totalSize`. This means callers who rely on stats from the method call may wrong results.

In particular, when `spark.sql.statistics.size.autoUpdate.enabled` is turned on, during insert overwrite Spark will first list partitions and get old statistics, and then compare them with new statistics and see which partitions need to be updated. This issue will sometimes cause it to update all partitions instead of only those partitions that have been touched.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added a new test.

### CODE REVIEW

**Comment by sunchao** [comment]
cc @viirya @dongjoon-hyun @cloud-fan

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @sunchao . Will take a look as soon as possible after finishing the ongoing work.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 1]
```diff
@@ -1290,6 +1291,7 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
         res.filter(p => isPartialPartitionSpec(spec, toMetaStorePartitionSpec(p.spec)))
       case _ => res
     }
+    parts.map(restorePartitionMetadata(_, catalogTable))
```
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:1294]: Actaully, `getPartition` does `restorePartitionMetadata` as expected. But `listPartitions` doesn't follow this behavior.

Comment (viirya) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:1294]: BTW, `restorePartitionMetadata ` is added by SPARK-21213. From its description, seems at the moment, the patch proposed to analyze table partition to collect partition-level statistics.

Comment (viirya) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:1294]: So I guess it was not considered for the case of listing partitions.

**Review by viirya — COMMENTED** [review]
Do we also need to update other partition related method like `listPartitionsByFilter`?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala (hunk 1 above)
Comment (sunchao) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala:1294]: Yes looks like not all operations are updated with `restorePartitionMetadata`

**Comment by sunchao** [comment]
> Do we also need to update other partition related method like listPartitionsByFilter?

Yes good point. Let me update `listPartitionsByFilter` too

**Review by viirya — APPROVED** [review]

**Comment by sunchao** [comment]
Merged to master/branch-3.5/branch-3.4, thanks!

**Comment by wangyum** [comment]
LGTM.

**Comment by cloud-fan** [comment]
late LGTM

**Review by dongjoon-hyun — COMMENTED** [review]
+1, late LGTM. Thank you all.

**Pull request closed without merging**